exchangeBridge/exchange.go

```diff
@@ -0,0 +1,161 @@
+// does not support streaming.
+type ExchangeBridge interface {
+	GetEvent(ctx context.Context) (*eventapi.Envelope, error)
+}
+
+type exchangeBridgeService struct {
+	eventChan <-chan *events.Envelope
+	errChan   <-chan error
+}
+
+// NewExchangeBridgeService returns a server-side implementation of the ExchangeBridge interface. Given an existing
+// exchange, it will subscribe to that exchange with no filters. Each GetEvent call pops and returns an event buffered
+// in the subscription's channel of published events.
+func NewExchangeBridgeService(ctx context.Context, eventSource events.Subscriber) ExchangeBridge {
+	eventChan, errChan := eventSource.Subscribe(ctx)
+	return &exchangeBridgeService{
+		eventChan: eventChan,
+		errChan:   errChan,
+	}
+}
+
+func RegisterExchangeBridgeService(srv *ttrpc.Server, svc ExchangeBridge) {
+	srv.Register(ExchangeBridgeServiceName, map[string]ttrpc.Method{
+		GetEventMethodName: func(ctx context.Context, unmarshal func(interface{}) error) (interface{}, error) {
+			return svc.GetEvent(ctx)
+		},
+	})
+}
+
+func (s *exchangeBridgeService) GetEvent(ctx context.Context) (*eventapi.Envelope, error) {
+	var (
+		envelope *eventapi.Envelope
+		err error
+	)
+
+	select {
+	case receivedEnvelope := <-s.eventChan:
+		envelope = &eventapi.Envelope{
+			Timestamp: receivedEnvelope.Timestamp,
+			Namespace: receivedEnvelope.Namespace,
+			Topic:     receivedEnvelope.Topic,
+			Event:     receivedEnvelope.Event,
+		}
+	case err = <-s.errChan:
+	case <-ctx.Done():
+		if ctxErr := ctx.Err(); ctxErr != context.Canceled {
+			err = ctx.Err()
+		}
+	}
+
+	return envelope, err
+}
+
+type exchangeBridgeClient struct {
+	rpcClient *ttrpc.Client
+}
+
+func NewExchangeBridgeClient(rpcClient *ttrpc.Client) ExchangeBridge {
+	return &exchangeBridgeClient{
+		rpcClient: rpcClient,
+	}
+}
+
+func (c *exchangeBridgeClient) GetEvent(ctx context.Context) (*eventapi.Envelope, error) {
+	var req types.Empty
+	var resp eventapi.Envelope
+	if err := c.rpcClient.Call(ctx, ExchangeBridgeServiceName, GetEventMethodName, &req, &resp); err != nil {
+		return nil, err
+	}
+
+	return &resp, nil
+}
+
+// AttachExchangeBridge takes an existing ExchangeBridge and forwards each event retrieved over that bridge to the
+// provided forwarder. For example, if the ExchangeBridge is a client for a remote exchange and forwarder is a local
+// exchange, AttachExchangeBridge will result in each event published on the remote exchange to be forwarded to the
+// local exchange, essentially bridging the two together.
+func AttachExchangeBridge(ctx context.Context, exchangeBridge ExchangeBridge, forwarder events.Forwarder) error {
+	for {
+		select {
+		case <-ctx.Done():
+			if ctxErr := ctx.Err(); ctxErr != context.Canceled {
+				return ctxErr
+			} else {
+				return nil
+			}
+		default:
+		}
+
+		envelope, err := exchangeBridge.GetEvent(ctx)
+		if err != nil {
+			return err
+		}
```

Comment:
What happens if there are no new events published? Will it fail here with timeout and will stop pulling events?

Reply:
Looking through the TTRPC codebase I wasn't able to find any implementation of timeouts, so as far as I can tell the request will just block indefinitely until events are available or the connection is closed by either side. I also manually let a container sit around for 10 minutes and verified the task exit event was still propagated as expected afterwards (admittedly not the most robust test, but not sure how else to verify at this moment).

Let me know if I'm missing something.

Reply:
Yep. Timeouts are not yet implemented https://github.com/containerd/ttrpc/issues/3

Reply:
Thanks, that's good to know! Hopefully they add support for timeouts in a backwards-compatible way, but I'll assume they won't and see what I can do to cover this when adding tests (maybe just have a container sit around for 30ish seconds before existing or something like that).